src/main/java/com/google/api/codegen/config/SampleConfig.java

```diff
@@ -50,14 +99,110 @@ public static Builder newBuilder() {
+      } else {
+        sampleSpecs.add(spec);
+      }
+    }
+
+    // Then, check user specified sample IDs do not clash
+    Set<String> distinctIds = new HashSet<>();
+    Set<String> duplicateIds =
+        sampleSpecs
+            .stream()
+            .map(s -> s.getId())
+            .filter(id -> !id.isEmpty())
+            .filter(s -> !distinctIds.add(s))
+            .collect(Collectors.toSet());
+    Preconditions.checkArgument(
+        duplicateIds.isEmpty(),
+        "Found duplicate IDs: %s",
+        duplicateIds.stream().collect(Collectors.joining(", ")));
+
+    // Next, flat out calling forms
+    List<SampleSpecProto> flattenedSampleSpecs = new ArrayList<>();
+    for (SampleSpecProto spec : sampleSpecs) {
+      for (String pattern : spec.getCallingPatternsList()) {
+        sampleSpecs.add(spec.toBuilder().addCallingPatterns(pattern).build());
+      }
+    }
+
+    // We cannot auto-generate IDs for samples without user provided ones at this step.
+    // We need to further flat out calling forms at generation time because
+    // calling forms support regular expressions, and a single `calling_pattern`
+    // string may map to multiple calling forms.
+
+    // Construct `SampleConfig` objects.
+    return flattenedSampleSpecs
+        .stream()
+        .map(spec -> create(spec, interfaceConfigMap))
+        .collect(ImmutableList.toImmutableList());
+  }
+
+  private static SampleConfig create(
+      SampleSpecProto sampleSpec, Map<String, InterfaceConfig> interfaceConfigMap) {
+    InterfaceConfig interfaceConfig = interfaceConfigMap.get(sampleSpec.getService());
+    Preconditions.checkNotNull(
+        interfaceConfig, "can't find interface named %s", sampleSpec.getService());
+    Preconditions.checkState(
+        interfaceConfig instanceof GapicInterfaceConfig,
+        "can't generate samples for non-gapic libraries");
+
+    GapicInterfaceConfig gapicInterfaceConfig = (GapicInterfaceConfig) interfaceConfig;
+    MethodConfig methodConfig = gapicInterfaceConfig.getMethodConfigMap().get(sampleSpec.getRpc());
+    Preconditions.checkNotNull(methodConfig, "can't find method named %s", sampleSpec.getRpc());
+
+    return SampleConfig.newBuilder()
+        .id(sampleSpec.getId())
+        .title(sampleSpec.getTitle())
+        .description(sampleSpec.getDescription())
+        .interfaceConfig(gapicInterfaceConfig)
+        .methodConfig(methodConfig)
+        .requestConfigs(
+            ImmutableList.<RequestFieldProto>builder().addAll(sampleSpec.getRequestList()).build())
+        .responseConfigs(
+            ImmutableList.<ResponseStatementProto>builder()
+                .addAll(sampleSpec.getResponseList())
+                .build())
+        .regionTag(sampleSpec.getRegionTag())
+        .callingPattern(
+            sampleSpec.getCallingPatternsCount() > 0
+                ? sampleSpec.getCallingPatternsList().get(0)
+                : "")
```

Comment:
How about specifying `"default"` here, and everything downstream knows to use the explicitly named patterns? (Ie the inference happens here). Or even better, in the method above, as per my comment there.

Reply:
IIRC We have always been using empty string to represent a default calling form. Is it OK to support `default` later?

Reply:
We can support it later.